# **Importing Libraries**

In [1]:
#Importing required libraries
import pandas as pd
import numpy as np


# **Reading Files**

In [2]:
#Reading csv files
df1=pd.read_csv('/content/olist_customers_dataset.csv.zip')
df2=pd.read_csv('/content/olist_order_items_dataset.csv.zip')
df3=pd.read_csv('/content/olist_order_payments_dataset.csv.zip')
df4=pd.read_csv('/content/olist_order_reviews_dataset.csv')
df5=pd.read_csv('/content/olist_orders_dataset.csv.zip')
df6=pd.read_csv('/content/olist_products_dataset.csv.zip')
df7=pd.read_csv('/content/olist_sellers_dataset.csv')
df8=pd.read_csv('/content/product_category_name_translation.csv')

FileNotFoundError: [Errno 2] No such file or directory: '/content/olist_customers_dataset.csv.zip'

In [ ]:
#Columns in different dataframes
for name, df in zip(['customers_dataset','order_items_dataset', 'order_payments_dataset', 'order_reviews_dataset', 'orders_dataset.csv', 'products_dataset', 'sellers_dataset', "product_category_name_translation"], [df1, df2, df3, df4, df5, df6, df7, df8]):
    print(f"--- columns from {name} ---")
    display(df.columns)
    print("\n")

--- columns from customers_dataset ---


Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='object')



--- columns from order_items_dataset ---


Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='object')



--- columns from order_payments_dataset ---


Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='object')



--- columns from order_reviews_dataset ---


Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='object')



--- columns from orders_dataset.csv ---


Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')



--- columns from products_dataset ---


Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='object')



--- columns from sellers_dataset ---


Index(['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state'], dtype='object')



--- columns from product_category_name_translation ---


Index(['product_category_name', 'product_category_name_english'], dtype='object')

In [ ]:
#Displaying different csv files
for name, df in zip(['customers_dataset', 'order_items_dataset', 'order_payments_dataset', 'order_reviews_dataset', 'orders_dataset.csv', 'products_dataset', 'sellers_dataset', "product_category_name_translation"], [df1, df2, df3, df4, df5, df6, df7, df8]):
    print(f"--- Sample from {name} ---")
    display(df.sample(5))
    print("\n")

--- Sample from customers_dataset ---


customer_id                customer_unique_id  \
69209  dd016c21610e21f5aea27c2e560d1b76  980331124f1c9bb761311867a6d9dc7d   
12145  c1da9b0b8b59fe31a48c170dddc2cd98  a6e92e646f11fbc93be77b8ac42c3f9b   
88768  75b8fb170128aac9596ec842cf844c23  3bbad45a834fd8a6f68ce54460a53f55   
79119  b03ca09111fbab1d3dab28a28413356c  ecc6b07c8e4cbe145c0119b3a10a3bde   
86851  07d32200708d263fc5006988e4b6eff7  0b08d01a339342b612080ef1cb877fd6   

       customer_zip_code_prefix       customer_city customer_state  
69209                     60740           fortaleza             CE  
12145                     59120               natal             RN  
88768                     37558        pouso alegre             MG  
79119                     78530  peixoto de azevedo             MT  
86851                     89070            blumenau             SC



--- Sample from order_items_dataset ---


order_id  order_item_id  \
82682  bbccc293dde3a6b72a942c4117678017              1   
84386  bfb88796b657bb5483be15669e9f3496              1   
9386   1573c1514d9e6b02151d3f7a1e9b0b20              2   
62769  8f505b97d690e877f34131b13f828667              2   
56180  7fc9cfbe25b5e0d747a0b4ca5678aaf3              1   

                             product_id                         seller_id  \
82682  6eb70dfb283dc3f645ed7b03c990dc72  7d13fca15225358621be4086e1eb0964   
84386  eb9b44e05684527fbfdfd0ff5cb86250  df0f42bc4c2142eacf0eaf2cffd0cfbb   
9386   a87567bdce60de6f901e3d7c6d8df820  318f287a62ab7ac10b703ac37435a231   
62769  f434709eecc62585cc43f03ef042ae97  8bb48dc19fccaa8613b6229bf7f452a2   
56180  7a10781637204d8d10485c71a6108a2e  4869f7a5dfa277a7dca6462dcf3b52b2   

       shipping_limit_date  price  freight_value  
82682  2018-05-22 09:30:58   55.0          18.27  
84386  2018-03-08 15:15:39  104.0          15.48  
9386   2018-02-21 16:46:56   74.0          13.08  
62769  2017-11-23 04:15:34   20.4          16.11  
56180  2018-03-29 20:31:46  219.0          24.11



--- Sample from order_payments_dataset ---


order_id  payment_sequential payment_type  \
33929  bb27293f7b2e1c705bf345fd57ecd0dd                   1  credit_card   
40902  93a0e7a8ac210479d99b1726f697651c                   1  credit_card   
96441  6d34c087b843ae0fdb296f15d948e45f                   3      voucher   
48832  33097b1461e9e4dce6383085eef494b9                   1  credit_card   
67398  ab3bc480a02b5a058759d5855428c22f                   1  credit_card   

       payment_installments  payment_value  
33929                     1         199.42  
40902                     6         905.76  
96441                     1         200.00  
48832                     1          68.22  
67398                     1          19.36



--- Sample from order_reviews_dataset ---


review_id                          order_id  \
93538  e57583df6b569e14aa1b0da99317bcbe  e1df06d2e0fcfed2ec12f002d54da4f0   
63249  290da3d3a680aac8c788949dc1a5b936  7106d4481a2ba708b934d912e3b4413a   
36454  a77fc422f3192cbbcca7c5a0f8d6d80a  5d8b449936e528d0d68ca11c6e42a224   
47483  822f239dac4cfd170caed88728ae4f51  c8d07bf0161cd0a2aacc163c238f397e   
64439  53146efec2baa456d82912698ebb1816  bd76688da0406f5c377080cc43888b5e   

       review_score review_comment_title  \
93538             5                  NaN   
63249             3                  NaN   
36454             5                  NaN   
47483             5                  NaN   
64439             5                  NaN   

                            review_comment_message review_creation_date  \
93538  Adorei o tapete. Felpudo com bom acabamento  2018-06-30 00:00:00   
63249                                          NaN  2018-02-17 00:00:00   
36454                                          NaN  2017-09-07 00:00:00   
47483             Produto entregue entes do prazo.  2017-07-14 00:00:00   
64439                                          NaN  2017-07-08 00:00:00   

      review_answer_timestamp  
93538     2018-07-07 10:04:39  
63249     2018-02-17 22:46:27  
36454     2017-09-08 17:02:33  
47483     2017-07-16 23:28:41  
64439     2017-07-10 23:11:23



--- Sample from orders_dataset.csv ---


order_id                       customer_id  \
68775  9e766b8cbac373f4795d99c05f57514c  e46281bc93a3e0a5e9043f4a19dd541c   
21713  3e4a2b96719ecae342e5d413e97a1be2  a93e6e1bad3b9d95239b7dd652a1ff13   
60661  92e0a2b039d9ce627cbcc94ecf879f87  1ff8a1ae692a49c3204aa327a2c9cc7f   
89238  89ae9eea43fd5c68102875e05af1c2da  2ef9528dba475177acedbdfe80029ba1   
85405  231d32fab33b75ec36a075c7982d0803  eb31a8edc763589acbb30bcf919d9c91   

      order_status order_purchase_timestamp    order_approved_at  \
68775    delivered      2018-07-08 19:40:11  2018-07-08 19:50:14   
21713    delivered      2018-02-27 11:06:32  2018-02-27 11:15:32   
60661    delivered      2017-11-13 15:19:44  2017-11-13 15:31:35   
89238    delivered      2017-08-29 20:42:57  2017-08-29 20:50:22   
85405    delivered      2018-02-21 12:46:05  2018-02-21 13:08:55   

      order_delivered_carrier_date order_delivered_customer_date  \
68775          2018-07-10 14:54:00           2018-07-19 20:58:45   
21713          2018-03-12 18:22:25           2018-03-21 22:38:25   
60661          2017-11-14 19:58:58           2017-11-22 21:56:49   
89238          2017-09-05 22:47:58           2017-09-12 20:25:27   
85405          2018-02-22 18:54:17           2018-03-10 17:28:40   

      order_estimated_delivery_date  
68775           2018-08-06 00:00:00  
21713           2018-03-21 00:00:00  
60661           2017-12-01 00:00:00  
89238           2017-09-19 00:00:00  
85405           2018-03-13 00:00:00



--- Sample from products_dataset ---


product_id  product_category_name  \
14283  075b253cd6c59b821c1dab51c1ae56b9     ferramentas_jardim   
10075  c224f464aeeb2c6af33f0682a181efa7  utilidades_domesticas   
9045   44efeca9e2a12a4e5af1940d1d5d7c18               pet_shop   
28507  2a71535169f1f7a56b266e6877bb8cf9              telefonia   
32032  14d617ec6c1f7a295007b14080e84a97     ferramentas_jardim   

       product_name_lenght  product_description_lenght  product_photos_qty  \
14283                 53.0                       768.0                 1.0   
10075                 60.0                      1809.0                 1.0   
9045                  59.0                       590.0                 5.0   
28507                 22.0                       285.0                 3.0   
32032                 58.0                       934.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
14283             350.0               20.0                5.0   
10075           12600.0               16.0                2.0   
9045              500.0               47.0                6.0   
28507             100.0               19.0                3.0   
32032             500.0               16.0               28.0   

       product_width_cm  
14283              20.0  
10075              11.0  
9045               36.0  
28507              11.0  
32032              11.0



--- Sample from sellers_dataset ---


seller_id  seller_zip_code_prefix  seller_city  \
504   62c50c1af4dfdc4149d25c5222043d39                    4782    sao paulo   
1064  a154d7316f158bb42e6fa18bbe3afd3a                    3804    sao paulo   
653   3364a91ec4d56c98e44174de954b94f6                   76804  porto velho   
2449  134b867a0f7f8bbe6abf10f91874891e                   13272     valinhos   
503   80cfd480be3eb6d6267139f5cb793555                    8223    sao paulo   

     seller_state  
504            SP  
1064           SP  
653            RO  
2449           SP  
503            SP



--- Sample from product_category_name_translation ---


product_category_name    product_category_name_english
25  construcao_ferramentas_construcao  construction_tools_construction
18                    eletroportateis                 small_appliances
46                              artes                              art
59                          cine_foto                       cine_photo
64                  cds_dvds_musicais                cds_dvds_musicals

# **Merging files**

In [ ]:
#Merging all columns of all files(except Geolocation file)
orders_customers = pd.merge(df5, df1, on='customer_id', how='left')
order_items_products = pd.merge(df2, df6, on='product_id', how='left')
order_items_products_sellers = pd.merge(order_items_products, df7, on='seller_id', how='left')
orders_items_products_sellers_customers = pd.merge(orders_customers, order_items_products_sellers, on='order_id', how='left')
merged_df = pd.merge(orders_items_products_sellers_customers, df4, on='order_id', how='left')
merged_df = pd.merge(merged_df, df5, on='order_id', how='left')

merged_df = pd.merge(merged_df, df8, on='product_category_name', how='left')

# Geolocation data not merged

display(merged_df.shape)
display(merged_df.columns)
display(merged_df.sample(5))

(114092, 43)

Index(['order_id', 'customer_id_x', 'order_status_x',
       'order_purchase_timestamp_x', 'order_approved_at_x',
       'order_delivered_carrier_date_x', 'order_delivered_customer_date_x',
       'order_estimated_delivery_date_x', 'customer_unique_id',
       'customer_zip_code_prefix', 'customer_city', 'customer_state',
       'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date',
       'price', 'freight_value', 'product_category_name',
       'product_name_lenght', 'product_description_lenght',
       'product_photos_qty', 'product_weight_g', 'product_length_cm',
       'product_height_cm', 'product_width_cm', 'seller_zip_code_prefix',
       'seller_city', 'seller_state', 'review_id', 'review_score',
       'review_comment_title', 'review_comment_message',
       'review_creation_date', 'review_answer_timestamp', 'customer_id_y',
       'order_status_y', 'order_purchase_timestamp_y', 'order_approved_at_y',
       'order_delivered_carrier_date_y', 'order_delivered_custo

order_id                     customer_id_x  \
93142  8654b9ba7d0240807fe0ecdab1f54d5c  eb63cbbc3edd0096ac53f7fe81533f22   
62528  2fc58fd97ab3288872ad6b35bbd712c3  85fdc9f140218413947b78ea86ce9855   
59005  3f5dec4a044b171bbe8c2313102c045b  f35aff4ab7e91c2c5b2ae732007777fe   
8914   09903d25ea3d162aeb40057420caa5bb  e4caff66029e8290ad28fab525631c74   
57825  93240f6a90ae9889c692787235e66a0c  0566ea068bbe90dca2a0204179e22f37   

      order_status_x order_purchase_timestamp_x  order_approved_at_x  \
93142      delivered        2017-12-18 23:24:30  2017-12-19 01:31:25   
62528      delivered        2018-04-29 09:34:52  2018-04-29 09:55:01   
59005      delivered        2017-08-31 09:50:09  2017-09-02 02:50:19   
8914       delivered        2017-11-24 13:26:15  2017-11-24 15:34:54   
57825      delivered        2018-01-11 11:16:49  2018-01-13 04:55:10   

      order_delivered_carrier_date_x order_delivered_customer_date_x  \
93142            2017-12-20 20:04:15             2018-01-08 16:15:52   
62528            2018-04-30 15:05:00             2018-05-02 21:16:52   
59005            2017-09-08 18:32:36             2017-09-15 21:27:37   
8914             2017-11-27 11:13:13             2017-12-26 21:33:40   
57825            2018-01-15 20:16:38             2018-01-23 14:17:22   

      order_estimated_delivery_date_x                customer_unique_id  \
93142             2018-01-22 00:00:00  2e4650d71b07066c339dc2544826c7b9   
62528             2018-05-09 00:00:00  00539155ce68fbe677934aebdb17e09b   
59005             2017-09-13 00:00:00  620cb4f374cfbb348e54fdb28a84d115   
8914              2017-12-21 00:00:00  e6bce3bae4ce9bba358e56e51d9e005a   
57825             2018-02-15 00:00:00  1da09dd64e235e7c2f29a4faff33535c   

       customer_zip_code_prefix  ... review_creation_date  \
93142                     78365  ...  2018-01-09 00:00:00   
62528                      8290  ...  2018-05-03 00:00:00   
59005                      5206  ...  2017-09-16 00:00:00   
8914                      60721  ...  2017-12-22 00:00:00   
57825                     24020  ...  2018-01-24 00:00:00   

      review_answer_timestamp                     customer_id_y  \
93142     2018-01-09 19:27:53  eb63cbbc3edd0096ac53f7fe81533f22   
62528     2018-05-04 13:03:44  85fdc9f140218413947b78ea86ce9855   
59005     2017-09-19 08:43:31  f35aff4ab7e91c2c5b2ae732007777fe   
8914      2017-12-22 15:55:24  e4caff66029e8290ad28fab525631c74   
57825     2018-01-26 17:40:04  0566ea068bbe90dca2a0204179e22f37   

      order_status_y order_purchase_timestamp_y  order_approved_at_y  \
93142      delivered        2017-12-18 23:24:30  2017-12-19 01:31:25   
62528      delivered        2018-04-29 09:34:52  2018-04-29 09:55:01   
59005      delivered        2017-08-31 09:50:09  2017-09-02 02:50:19   
8914       delivered        2017-11-24 13:26:15  2017-11-24 15:34:54   
57825      delivered        2018-01-11 11:16:49  2018-01-13 04:55:10   

       order_delivered_carrier_date_y  order_delivered_customer_date_y  \
93142             2017-12-20 20:04:15              2018-01-08 16:15:52   
62528             2018-04-30 15:05:00              2018-05-02 21:16:52   
59005             2017-09-08 18:32:36              2017-09-15 21:27:37   
8914              2017-11-27 11:13:13              2017-12-26 21:33:40   
57825             2018-01-15 20:16:38              2018-01-23 14:17:22   

      order_estimated_delivery_date_y  product_category_name_english  
93142             2018-01-22 00:00:00                 sports_leisure  
62528             2018-05-09 00:00:00                furniture_decor  
59005             2017-09-13 00:00:00                 sports_leisure  
8914              2017-12-21 00:00:00                  watches_gifts  
57825             2018-02-15 00:00:00          computers_accessories  

[5 rows x 43 columns]

In [ ]:
#Merging specific columns of different files
order_data = df5.merge(df2, on='order_id') \
                   .merge(df3, on='order_id') \
                   .merge(df1, on='customer_id') \
                   .merge(df4[['order_id', 'review_score']], on='order_id', how='left') \
                   .merge(df6[['product_id', 'product_category_name']], on='product_id', how='left') \
                   .merge(df8, on='product_category_name', how='left')


# Geolocation data not merged

order_data = order_data.drop('product_category_name', axis=1) #dropping Portuguese product name(keeping English name)

#displaying
display(order_data.shape)
display(order_data.columns)
display(order_data.sample(5))

#order details(id,status,purchase date, approved date, delivery dates,product id, shipping date,price,freight value)
#customer details(id,zip_code,city,state),seller(id,) product etails(id,category_name)
#payement details(payment type,installments,vaue,sequential)

(118307, 24)

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date',
       'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date',
       'price', 'freight_value', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value', 'customer_unique_id',
       'customer_zip_code_prefix', 'customer_city', 'customer_state',
       'review_score', 'product_category_name_english'],
      dtype='object')

order_id                       customer_id  \
118162  e9e64a17afa9653aacf2616d94c005b8  b4cd0522e632e481f8eaf766a2646e86   
55144   14f478070db62f51c33367debce4f1d1  fed7a1c4146b45f0ddb2b285da5379cf   
10695   142c07328adba2427b3029a2b5a1f503  bb57526f1e8fe74f92c4c7e4ced09303   
45048   6413a8cd79f9720426fd12968b388228  315d30072434c1feda91ba8676b54a25   
36410   1d5f437b878f76103ac129b13e81bbbb  e17feb0f580f7f763ccb5684dd52cb1e   

       order_status order_purchase_timestamp    order_approved_at  \
118162   processing      2018-01-05 23:07:24  2018-01-09 07:18:05   
55144     delivered      2017-07-04 20:32:22  2017-07-05 17:38:28   
10695     delivered      2018-01-16 10:48:40  2018-01-16 11:10:53   
45048     delivered      2018-06-25 10:38:46  2018-06-25 10:53:51   
36410     delivered      2018-08-03 13:19:41  2018-08-04 05:05:26   

       order_delivered_carrier_date order_delivered_customer_date  \
118162                          NaN                           NaN   
55144           2017-07-06 15:17:54           2017-07-12 20:42:56   
10695           2018-01-17 15:48:07           2018-01-23 18:17:05   
45048           2018-06-29 12:27:00           2018-07-03 16:48:51   
36410           2018-08-13 07:35:00           2018-08-14 17:46:31   

       order_estimated_delivery_date  order_item_id  \
118162           2018-02-06 00:00:00              1   
55144            2017-07-26 00:00:00              1   
10695            2018-02-06 00:00:00              1   
45048            2018-07-31 00:00:00              1   
36410            2018-08-17 00:00:00              2   

                              product_id  ... payment_sequential payment_type  \
118162  0a57f7d2c983bcf8188589a5fea4a8da  ...                  1       boleto   
55144   473795a355d29305c3ea6b156833adf5  ...                  1  credit_card   
10695   d692d19cee022f7acab76373f297ad6a  ...                  1  credit_card   
45048   d016a1a4d0898c213be8443b0c99c784  ...                  1  credit_card   
36410   ee300c0281d54f57ed53d622f7e8dbff  ...                  1       boleto   

        payment_installments  payment_value                customer_unique_id  \
118162                     1         144.56  965191786b70912f793e4a777fd623cd   
55144                      3          72.05  7c5ccf4e7f7a4ab06c45c27d038e8c10   
10695                      5         563.64  d040aa4ef4eee97186650ee10600e77a   
45048                      6         889.28  22eecffbc2e4e56747eb329d96051962   
36410                      1         138.14  881483770fdbfedae83075ad3ac0a289   

       customer_zip_code_prefix  customer_city  customer_state review_score  \
118162                    13050       campinas              SP          1.0   
55144                      5508      sao paulo              SP          4.0   
10695                     14340      brodowski              SP          5.0   
45048                     24030        niteroi              RJ          1.0   
36410                     13214        jundiai              SP          1.0   

        product_category_name_english  
118162                  watches_gifts  
55144                       perfumery  
10695                  bed_bath_table  
45048           computers_accessories  
36410                air_conditioning  

[5 rows x 24 columns]

In [ ]:
#those col names which are not in order_data but there in merged_data
set(merged_df.columns) - set(order_data.columns)

#check if any of these are required columns

## **EDA**

In [ ]:
order_data.info()

In [ ]:
order_data.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                  15
order_delivered_carrier_date     1254
order_delivered_customer_date    2588
order_estimated_delivery_date       0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
payment_sequential                  0
payment_type                        0
payment_installments                0
payment_value                       0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
review_score                      978
product_category_name_english    1734
dtype: int64

#We need to remove these null rows for required columns, some columns can be skipped altogether

In [ ]:
#Converting data type of date col from object to datetime
date_columns = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date',
    'shipping_limit_date',
]

for col in date_columns:
    order_data[col] = pd.to_datetime(order_data[col], errors='coerce')

order_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 118307 entries, 0 to 118306
Data columns (total 24 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       118307 non-null  object        
 1   customer_id                    118307 non-null  object        
 2   order_status                   118307 non-null  object        
 3   order_purchase_timestamp       118307 non-null  datetime64[ns]
 4   order_approved_at              118292 non-null  datetime64[ns]
 5   order_delivered_carrier_date   117053 non-null  datetime64[ns]
 6   order_delivered_customer_date  115719 non-null  datetime64[ns]
 7   order_estimated_delivery_date  118307 non-null  datetime64[ns]
 8   order_item_id                  118307 non-null  int64         
 9   product_id                     118307 non-null  object        
 10  seller_id                      118307 non-null  object        
 11  

In [ ]:
#differentiating col as categorical, numerical, datetime columns
categorical_col = [col for col in order_data.columns if order_data[col].dtype == 'object']
datetime_col = [col for col in order_data.columns if order_data[col].dtype == 'datetime64[ns]']
numerical_col = [col for col in order_data.columns if order_data[col].dtype != 'object' and order_data[col].dtype != 'datetime64[ns]']

Need to check if some col seems like numerical/categorical but are categorical/numerical

In [ ]:
order_data[categorical_col].describe()

order_id                       customer_id  \
count                             118307                            118307   
unique                             98665                             98665   
top     895ab968e7bb0d5659d16cd74cd1650c  270c23a11d024a44c896d1894b261a83   
freq                                  63                                63   

       order_status                        product_id  \
count        118307                            118307   
unique            7                             32951   
top       delivered  aca2eb7d00ea1a7b8ebd4e68314663af   
freq         115720                               536   

                               seller_id payment_type  \
count                             118307       118307   
unique                              3095            4   
top     4a3ca9315b744ce9f8e9374361493884  credit_card   
freq                                2155        87258   

                      customer_unique_id customer_city customer_state  \
count                             118307        118307         118307   
unique                             95419          4110             27   
top     9a736b248f67d166d2fbb006bcb877c3     sao paulo             SP   
freq                                  75         18727          49862   

       product_category_name_english  
count                         116573  
unique                            71  
top                   bed_bath_table  
freq                           11988

In [ ]:
order_data[numerical_col].describe()

order_item_id          price  freight_value  payment_sequential  \
count  118307.000000  118307.000000  118307.000000       118307.000000   
mean        1.196523     120.648522      20.032823            1.094069   
std         0.699474     184.111631      15.836814            0.728579   
min         1.000000       0.850000       0.000000            1.000000   
25%         1.000000      39.900000      13.080000            1.000000   
50%         1.000000      74.900000      16.280000            1.000000   
75%         1.000000     134.900000      21.180000            1.000000   
max        21.000000    6735.000000     409.680000           29.000000   

       payment_installments  payment_value  customer_zip_code_prefix  \
count         118307.000000  118307.000000             118307.000000   
mean               2.943317     172.575102              35051.177276   
std                2.777877     267.110727              29824.786247   
min                0.000000       0.000000               1003.000000   
25%                1.000000      60.850000              11310.000000   
50%                2.000000     108.200000              24310.000000   
75%                4.000000     189.260000              58640.000000   
max               24.000000   13664.080000              99990.000000   

        review_score  
count  117329.000000  
mean        4.031467  
std         1.387927  
min         1.000000  
25%         4.000000  
50%         5.000000  
75%         5.000000  
max         5.000000

In [ ]:
order_data[datetime_col].describe()

order_purchase_timestamp              order_approved_at  \
count                         118307                         118292   
mean   2017-12-30 06:48:50.040707584  2017-12-30 18:16:50.277744896   
min              2016-09-04 21:15:19            2016-10-04 09:43:32   
25%              2017-09-11 15:16:26            2017-09-12 04:25:26   
50%              2018-01-17 22:39:40     2018-01-18 10:24:29.500000   
75%              2018-05-03 16:34:57            2018-05-03 23:59:33   
max              2018-09-03 09:06:57            2018-09-03 17:40:06   

        order_delivered_carrier_date  order_delivered_customer_date  \
count                         117053                         115719   
mean   2018-01-03 08:45:19.290586112  2018-01-12 21:11:40.346200576   
min              2016-10-08 10:34:01            2016-10-11 13:46:32   
25%              2017-09-14 19:53:45            2017-09-22 22:10:19   
50%              2018-01-23 17:09:46            2018-02-01 10:57:52   
75%              2018-05-07 14:57:00     2018-05-15 00:13:14.500000   
max              2018-09-11 19:48:28            2018-10-17 13:22:46   

       order_estimated_delivery_date            shipping_limit_date  
count                         118307                         118307  
mean   2018-01-23 03:38:53.045381632  2018-01-05 22:23:29.509885184  
min              2016-10-20 00:00:00            2016-09-19 00:15:34  
25%              2017-10-02 00:00:00            2017-09-18 14:30:36  
50%              2018-02-14 00:00:00            2018-01-25 04:55:43  
75%              2018-05-25 00:00:00     2018-05-10 02:51:41.500000  
max              2018-10-25 00:00:00            2020-04-09 22:35:08

In [ ]:
order_data['order_status'].value_counts()

order_status
delivered      115720
shipped          1255
canceled          570
invoiced          376
processing        376
unavailable         7
approved            3
Name: count, dtype: int64

#Since we have to calculate CLV, cancelled, shipped, unavailable,processing orders are of not much use to us

In [ ]:
order_data.drop(order_data[order_data['order_status'].isin(['canceled', 'shipped', 'unavailable', 'processing'])].index, inplace=True)

In [ ]:
order_data['order_status'].value_counts()

order_status
delivered    115720
invoiced        376
approved          3
Name: count, dtype: int64

In [ ]:
order_data.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                  15
order_delivered_carrier_date      381
order_delivered_customer_date     387
order_estimated_delivery_date       0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
payment_sequential                  0
payment_type                        0
payment_installments                0
payment_value                       0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
review_score                      867
product_category_name_english    1673
dtype: int64

Earlier, missing values in order_delivered_carrier_date was 1254
& in order_delivered_customer_date was 2588 which have now reduced by significant amount